### PR TITLE
WIP (not tested): Command Filters API

### PR DESCRIPTION
**Pull request by nickrobson**
_TL;DR: Relatively minor revamp of how commands are created by API consumers._

**Example Usage**

```java
// Create a filter that passes case-insensitive "start" commands to a StartHandler
CommandFilter startCommand = new TextFilter("start", false, new StartHandler());

// Create a filter that passes case-insensitive "help" commands to a HelpHandler
CommandFilter helpCommand = new TextFilter("help", false, new HelpHandler());

// Create a filter that only allows commands in the form /command@mycustombot (or PM)
CommandFilter myCommands = new MentionFilter(startCommand, helpCommand);

// Registers these commands
bot.getCommandRegistry().registerCommand(myCommands);
```

---

**The current method of adding command handlers involves:**

1. Creating a CommandListener which tests if a Command will be handled.
2. Creating a CommandHandler which handles any Commands passed to it.
3. Registering these so only Commands allowed by the Listener are handled by the Handler.

---

**Why do I think the current system is convoluted?**

I personally do not see why listeners and handlers should be entirely separated in functionality. There is enough overlap between the two types where a handler could be a listener that when testing just executes the desired functionality and says "Yup! It passed me!"

---

**Information on the new command filters API**

This API merges the concepts of listeners and handlers into filters. Handlers should extend `CommandHandler`, and filters should extend `CommandFilter`, though this is not enforced. It is however _highly recommended_ that new filters add a constructor that has a super call providing children to the CommandFilter constructor: `CommandFilter(CommandFilter... children)`. This allows for layering.

I decided **not** to use `CommandFilter[]` as the parameter type to prevent horrifically verbose code such as `new CommandFilter(new CommandFilter[]{ new MentionFilter(), ... });`

It comes with three filter implementations out of the box: TextFilter (matches text directly), RegexFilter (matches a regular expression), and MentionFilter (was the bot mentioned?).

I believe that this API simplifies the creation of more complex command handlers as they can be layered, instead of reinventing the wheel each time more complex functionality is required.

As evident in the implementations of the three provided filters, it is easy to implement new filters - the longest implementation is the MentionFilter which is two lines of code split up for legibility, the others are one line each.

### CODE REVIEW

**Comment by bo0tzz** [comment]
+1

**Comment by mkotb** [comment]
Hey!

Things look pretty good, and I'd be happy to accept these changes. I just woke up, so I'll find time during the day today to look over the code and either approve it for merging or request changes. If possible, would you be willing to update the wiki if and when I accept this PR? Thanks!

File: jtelegrambotapi-core/src/main/java/com/jtelegram/api/commands/Command.java [hunk 1]
```diff
@@ -13,14 +13,10 @@
...
+    private final String[] args;
     private final TextMessage baseMessage;
 
     public String getArgsAsText() {
-        StringBuilder sb = new StringBuilder();
-
-        args.forEach(s->sb.append(s).append(" "));
-
-        return sb.toString().trim();
+        return String.join(" ", args);
```
Comment (aaomidi) [inline jtelegrambotapi-core/src/main/java/com/jtelegram/api/commands/Command.java]: We had a few reasons to use Lists over Arrays.

- We can force a list to be immutable and reuse.
- It is more OOP compared to Arrays.
- I can guarantee a List won't have a null element, not the same with an Array.
- I can use a List with various collection apis, and use streams without relying on a separate class.

File: jtelegrambotapi-core/src/main/java/com/jtelegram/api/commands/CommandHandler.java [hunk 2]
```diff
@@ -1,7 +1,20 @@
...
+import com.jtelegram.api.commands.filters.CommandFilter;
+import com.jtelegram.api.events.message.TextMessageEvent;
 
-    void onCommand(Command command);
+public abstract class CommandHandler extends CommandFilter {
+
+    public abstract void onCommand(TextMessageEvent event, Command command);
+
+    @Override
+    protected final boolean _test(TextMessageEvent event, Command command) {
```
Comment (aaomidi) [inline jtelegrambotapi-core/src/main/java/com/jtelegram/api/commands/CommandHandler.java]: Let's not use methods with underlines. AFAIK that is out of the java conventions.

File: jtelegrambotapi-core/src/main/java/com/jtelegram/api/commands/CommandRegistry.java [hunk 3]
```diff
@@ -13,48 +13,39 @@
...
-                baseCommand = baseCommand.substring(1);
-                break;
-            }
-        }
-
-        String args[] = messageText.split(" ");
-        List<String> argsList = new ArrayList<>(args.length > 1 ? args.length : 0);
+        List<MessageEntity> entities = message.getEntities();
+        MessageEntity first = !entities.isEmpty() ? entities.get(0) : null;
+        String baseCommand = first != null && first.getType() == MessageEntityType.BOT_COMMAND ? first.getContent().substring(1) : null;
```
Comment (aaomidi) [inline jtelegrambotapi-core/src/main/java/com/jtelegram/api/commands/CommandRegistry.java]: Open this up imo, it's annoying to read.

File: jtelegrambotapi-core/src/main/java/com/jtelegram/api/commands/filters/CommandFilter.java [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+ */
+public abstract class CommandFilter {
+
+    private final CommandFilter[] children;
+
+    public CommandFilter(CommandFilter... children) {
+        this.children = children;
+    }
+
+    protected abstract boolean _test(TextMessageEvent event, Command command);
```
Comment (aaomidi) [inline jtelegrambotapi-core/src/main/java/com/jtelegram/api/commands/filters/CommandFilter.java]: Same as above

File: jtelegrambotapi-core/src/main/java/com/jtelegram/api/commands/filters/CommandFilter.java [hunk 5]
```diff
@@ -0,0 +1,30 @@
+package com.jtelegram.api.commands.filters;
+
+import com.jtelegram.api.commands.Command;
+import com.jtelegram.api.events.message.TextMessageEvent;
+
+/**
+ * @author Nick Robson
```
Comment (aaomidi) [inline jtelegrambotapi-core/src/main/java/com/jtelegram/api/commands/filters/CommandFilter.java]: We're not allowing these, not to take away from your contributions but because this class will probably be modified by others in the future and this takes away from their contributions.

File: jtelegrambotapi-core/src/main/java/com/jtelegram/api/commands/CommandHandler.java (hunk 2 above)
Comment (mkotb) [reply jtelegrambotapi-core/src/main/java/com/jtelegram/api/commands/CommandHandler.java]:
```
Methods should be verbs, in mixed case with the first letter lowercase, with the first letter
of each internal word capitalized.
```

[Code Conventions for the Java Programming Language: 9. Naming Conventions](http://www.oracle.com/technetwork/java/codeconventions-135099.html)

File: jtelegrambotapi-core/src/main/java/com/jtelegram/api/commands/filters/MentionFilter.java [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+    }
+
+    @Override
+    protected boolean _test(TextMessageEvent event, Command command) {
+        String botUsername = event.getBot().getBotInfo().getUsername();
+
+        return event.getMessage().getChat().getType() == ChatType.PRIVATE
+                || event.getMessage().getEntities().stream()
+                        .filter(me -> me.getType() == MessageEntityType.MENTION)
+                        .filter(me -> botUsername.equalsIgnoreCase(me.toString()))
```
Comment (mkotb) [inline jtelegrambotapi-core/src/main/java/com/jtelegram/api/commands/filters/MentionFilter.java]: Do mentions of the bot part as a message entity when they are part of a command? Just want to verify here

File: jtelegrambotapi-core/src/main/java/com/jtelegram/api/commands/CommandRegistry.java [hunk 7]
```diff
@@ -1,60 +1,69 @@
...
 
-    public void registerCommand(CommandListener listener, CommandHandler handler) {
-        this.listeners.put(listener, handler);
+    public void registerCommand(CommandFilter filter, CommandFilter... filters) {
+        this.listeners.add(new RootFilter(filter, filters));
     }
 
-    public void registerCommand(String command, CommandHandler handler) {
-        this.registerCommand(new TextCommandListener(command, false), handler);
+    public void registerCommand(String command, CommandFilter filter, CommandFilter... filters) {
```
Comment (mkotb) [inline jtelegrambotapi-core/src/main/java/com/jtelegram/api/commands/CommandRegistry.java]: A) You cannot use a lambda for this (You have to instantiate CommandHandler which is a filter for some reason)
B) This overall design isn't that intuitive

We need to have this work by having a basis of a command (basic string expression) and a consumer, from there further filters like ChatType, argument length, or whatever can be put on top as additions. Having the handler be a "filter" is confusing and is what makes this new system difficult to use.

File: jtelegrambotapi-core/src/main/java/com/jtelegram/api/commands/filters/CommandFilter.java [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+    protected abstract boolean preTest(TextMessageEvent event, Command command);
+
+    /**
+     * Checks to see if the given event and command test to be {@code true}
+     * according to this filter.
+     *
+     * @param event The event to be tested
+     * @param command The command data executed in this event
+     *
+     * @return True iff this filter allows this event and command combination to
```
Comment (mkotb) [inline jtelegrambotapi-core/src/main/java/com/jtelegram/api/commands/filters/CommandFilter.java]: Most people would interpret this as a typo, although you informed me that this means "if and only if." Please expand

File: jtelegrambotapi-core/src/main/java/com/jtelegram/api/commands/filters/ChatTypeFilter.java [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.EnumSet;
+import java.util.Set;
+
+/**
+ * A {@link CommandFilter} testing which type of chat the
+ * command was executed in.
+ *
+ * @author Nick Robson
```
Comment (mkotb) [inline jtelegrambotapi-core/src/main/java/com/jtelegram/api/commands/filters/ChatTypeFilter.java]: use of `@author`

File: jtelegrambotapi-core/src/main/java/com/jtelegram/api/commands/CommandRegistry.java (hunk 7 above)
Comment (nickrobson) [reply jtelegrambotapi-core/src/main/java/com/jtelegram/api/commands/CommandRegistry.java]: Lambdas are fixed :)
The signature is `boolean test(TextMessageEvent, Command)`.
Or cast to `CommandHandler` and enjoy `void onCommand(TextMessageEvent, Command)`
